File: yh_divisor/scripts/yh_divisor_sub.py
'''선생님 풀이 '''
#!/usr/bin/python
# -*- coding: utf-8 -*-

import rospy
from yh_divisor.msg import yh_divisor_msg

def msgCallback(msg):    
    n = msg.data   
    """ 정수의 약수를 모두 출력 """
    s = ""

    for i in range(1, n+1):
        if n % i ==0:
            s += str(i)
            s += " "
    print(s)              
    

def listner():
    rospy.init_node("yh_star_sub", anonymous=True)
    rospy.Subscriber("yh_star_topic", yh_divisor_msg, msgCallback)

    rospy.spin()

if __name__ == "__main__":
    listner()

File: yh_divisor/src/yh_divisor_sub.cpp
#include "ros/ros.h"
#include "yh_divisor/yh_divisor_msg.h"


void msgCallback(const yh_divisor::yh_divisor_msg::ConstPtr& msg)
{
    
    int n = msg->data;
    for(int i = 1; i <= n; i++)
    {
        if (n % i == 0 ) 
        {
            printf("%d ", i);

        }
    }
    printf("\n");
    
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "yh_divisor_sub");
    ros::NodeHandle nh;

    ros::Subscriber sub =  nh.subscribe("yh_divisor_topic", 100, msgCallback);

    ros::spin();  //프로그램 종료 방지 대기

    return 0;

}
